### PR TITLE
fix: avoid repetitive labels for integer/year/yearmonth axes

### LINKED ISSUE #8867: Repetitive labels for integer / time unit
```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "description": "A simple bar chart with embedded data.",
  "data": {
    "values": [
      {"a": 0, "b": 28},
      {"a": 1, "b": 38},
      {"a": 2, "b": 35}
    ]
  },
  "mark": "bar",
  "encoding": {
    "x": {
      "field": "a", 
      "type": "quantitative",
      "axis": {"format": "d"}
    },
    "y": {"field": "b", "type": "quantitative"}
  }
}
```

image

```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "description": "Google's stock price over time.",
  "data": {"url": "data/stocks.csv"},
  "transform": [
    {"filter": "datum.symbol==='GOOG'"},
    {"filter": {"field": "date", "timeUnit": "year", "range": [2006, 2007]}}

  ],
  "mark": "line",
  "encoding": {
    "x": {"field": "date", "timeUnit": "year", "type": "temporal"},
    "y": {"aggregate": "average", "field": "price", "type": "quantitative"}
  }
}
```

image

```
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "description": "Google's stock price over time.",
  "data": {"url": "data/stocks.csv"},
  "transform": [
    {"filter": "datum.symbol==='GOOG'"},
    {"filter": {"field": "date", "timeUnit": "year", "equal": 2006}},
    {"filter": {"field": "date", "timeUnit": "month", "range": [1,5]}}
  ],
  "width": 500,
  "mark": "line",
  "encoding": {
    "x": {"field": "date", "timeUnit": "yearmonth", "type": "temporal"},
    "y": {"aggregate": "average", "field": "price", "type": "quantitative"}
  }
}
```

image

### PR DESCRIPTION
**Pull request by kanitw**
Fixes #8867

avoid repetitive labels for integer/year/yearmonth axes

### CODE REVIEW

File: src/compile/axis/properties.ts [hunk 1]
```diff
@@ -287,34 +287,62 @@ export function defaultTickCount({
...
+      if (timeUnit) {
+        const normalizedTimeUnit = normalizeTimeUnit(timeUnit);
+        const unit = normalizedTimeUnit?.unit;
+        if (contains(['month', 'hours', 'day', 'quarter'], unit)) {
+          return undefined;
+        } else if (unit === 'year') {
+          return minOfDefaultTickCountAndSpan(scaleName, defaultSignal, val => `year(${val})`);
+        } else if (unit === 'yearmonth') {
+          return minOfDefaultTickCountAndSpan(scaleName, defaultSignal, val => `(year(${val}) * 12 + month(${val}))`);
+        }
```
Comment (fandu-db) [inline src/compile/axis/properties.ts:322]: nit: Since we are doing this for `month`, why not also do it for `quarter`?

**Review by fandu-db — APPROVED** [review]
LGTM. Two nit comments

1. About `quarter`
2. The new behavior seems not captured in any of the test PNGs

**Review by domoritz — CHANGES_REQUESTED** [review]
This feels complicated. We should instead use tickMinStep as suggested in https://github.com/vega/vega/issues/1503. Ideally this was fixed at the Vega level.

**Comment by kanitw** [comment]
>  We should instead use tickMinStep as suggested in https://github.com/vega/vega/issues/1503.

Tick min step doesn't really work for time binned axis.  So I'm not sure if it's worth the complexity to implement the solution twice. 

See this [spec](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAVhDYA7EABoQAEzjQATjRyZ289AHE2bMoLgByCAAIImNlADWlnJqhxLTOJgDucOLMsATAAMQQBs-MFBAOyUiipImCiooAzqgmjxiZy2DhCUUBDEIAC+SpjqSLIQAGZs6ghoANqg1TSCmHDqGcoJDAiUEACeCExsggC8k2b6APIz+mYlCi1tHV3JIK1wgsrdCXBxOghwAKqydBmDcEhdShWyZAeojZGhCpFRALrFxZ9KyOp7BlhLIDko-FA2MoaA80KAAB5wzY0ba7dA9DqHGjHM4XdBXG6HQY4J4gDoIHB1JDpJRIeE0CBInQOACyMIAyh0cEi4PDnBkjnAABQYzzY4WRACMAEppZZ+JZBSL9krIkFZSUfkpBkikGQyOo+Jj0EhSBVHnEtjsMs4aK4iSSMgBHBiVHSJHSkTXFIA).


> Ideally this was fixed at the Vega level.

Conceptually, I agree.  But practically, until that happen (which I don't have time for and unlikely that someone gonna fix soon), I don't think we should block this PR on that.

**Comment by kanitw** [comment]
> The new behavior seems not captured in any of the test PNGs

It does -- see all the SVG changes in the 4th commit:

image

Note that some Vega spec may have diffs, but have no visual change since the cap rule still return the original default value.

**Comment by kanitw** [comment]
Replaced with https://github.com/vega/vega-lite/pull/8872

**Pull request closed without merging**